### PR TITLE
Flash attention

### PR DESCRIPTION
**Pull request by philschmid**

### CODE REVIEW

**Comment by tmm1** [comment]
hi, thanks for documenting this.

i'm curious if you were able to train successfully with peft + flash attention?

i keep seeing loss spikes after a few iterations

**Comment by tmm1** [comment]
i tried on another machine and it works fine there, so something else is going on. i will continue investigation in https://github.com/Dao-AILab/flash-attention/issues/311

**Comment by philschmid** [comment]
@tmm1 what machine have you used where it is not working?

**Comment by tmm1** [comment]
it is a local machine with 3090. I got it working by starting with a fresh CUDA 11.8 conda environment. the flash-attn tests were failing in the broken env.

File: training/utils/llama_patch.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    # Past Key value support
+    if past_key_value is not None:
+        # reuse k, v, self_attention
+        key_states = torch.cat([past_key_value[0], key_states], dim=2)
+        value_states = torch.cat([past_key_value[1], value_states], dim=2)
+
+    past_key_value = (key_states, value_states) if use_cache else None
+
+    # PEFT Int4 support
+    query_states, key_states, value_states = [x.to(torch.bfloat16) for x in [query_states, key_states, value_states]]
```
Comment (tmm1) [inline training/utils/llama_patch.py:66]: converting to bf16 per forward pass here is slow

it is preferable to remove this (and the following assert), and instead convert these parts of the model to bf16 upfront

see for example: https://github.com/OpenAccess-AI-Collective/axolotl/pull/336

Comment (philschmid) [reply training/utils/llama_patch.py:66]: Thank you @tmm1 for sharing. Let me check that.

Comment (philschmid) [reply training/utils/llama_patch.py:66]: Added a branch here https://github.com/philschmid/deep-learning-pytorch-huggingface/compare/main...improve-flash
and test -> 20% faster. Do you know if there are any implication regarding "norm" not being in fp32? The QLoRA paper cast them on purposely to fp32 ([REF](https://github.com/artidoro/qlora/blob/7f4e95a68dc076bea9b3a413d2b512eca6d004e5/qlora.py#L396)).

Comment (tmm1) [reply training/utils/llama_patch.py:66]: I don't know TBH. I vaguely remember some reports of instability using fp16, but I think bf16 works better.

I notice that it's prepare_for_kbit_training that which seems to swap them back to fp32

https://github.com/huggingface/peft/blob/main/src/peft/utils/other.py#L95
